File: Source/Samples/Jacobi.Vst.Samples.Delay.VB/PluginParameterFactory.vb
Imports Jacobi.Vst.Core
Imports Jacobi.Vst.Framework

Friend Class PluginParameterFactory
    Public Categories As VstParameterCategoryCollection = New VstParameterCategoryCollection()
    Public ParameterInfos As VstParameterInfoCollection = New VstParameterInfoCollection()

    Public Sub New()
        Dim paramCat As VstParameterCategory = New VstParameterCategory()
        paramCat.Name = "Delay"

        Categories.Add(paramCat)
    End Sub

    Public Sub CreateParameters(ByVal parameters As VstParameterCollection)
        For Each paramInfo As VstParameterInfo In ParameterInfos
            If Categories.Count > 0 And paramInfo.Category Is Nothing Then paramInfo.Category = Categories(0)

            Dim param As VstParameter = New VstParameter(paramInfo)

            parameters.Add(param)
        Next
    End Sub

End Class

File: Source.Core/Samples/Jacobi.Vst.Samples.Delay/PluginParameterFactory.cs
namespace Jacobi.Vst.Samples.Delay
{
    using Jacobi.Vst.Core;
    using Jacobi.Vst.Framework;
    
    /// <summary>
    /// This class manages all parameters used by the plugin.
    /// </summary>
    internal class PluginParameterFactory
    {
        // we bypass thread safety concerns for now...

        /// <summary>
        /// A collection of parameter categories.
        /// </summary>
        public VstParameterCategoryCollection Categories = new VstParameterCategoryCollection();
        /// <summary>
        /// A collection of parameter definitions.
        /// </summary>
        public VstParameterInfoCollection ParameterInfos = new VstParameterInfoCollection();

        /// <summary>
        /// Initializes the new instance.
        /// </summary>
        public PluginParameterFactory()
        {
            VstParameterCategory paramCat = new VstParameterCategory();
            paramCat.Name = "Delay";

            Categories.Add(paramCat);
        }

        /// <summary>
        /// Fills the <paramref name="parameters"/> collection with all parameters.
        /// </summary>
        /// <param name="parameters">Must not be null.</param>
        /// <remarks>A <see cref="VstParameter"/> instance is created and linked up for each
        /// <see cref="VstParameterInfo"/> instance found in the <see cref="ParameterInfos"/> collection.</remarks>
        public void CreateParameters(VstParameterCollection parameters)
        {
            foreach (VstParameterInfo paramInfo in ParameterInfos)
            {
                if (Categories.Count > 0 && paramInfo.Category == null)
                {
                    paramInfo.Category = Categories[0];
                }

                VstParameter param = new VstParameter(paramInfo);

                parameters.Add(param);
            }
        }
    }
}
